warehouse/forklift/legacy.py

```diff
@@ -1330,11 +1331,19 @@ def file_upload(request):
                         "Binary wheel '{filename}' has an unsupported "
                         "platform tag '{plat}'.".format(filename=filename, plat=plat),
                     )
+            # Extract .metadata file
+            # https://www.python.org/dev/peps/pep-0658/#specification
+            with zipfile.ZipFile(temporary_filename) as zfp:
+                metafile = wheel_info.group("namever") + ".dist-info/METADATA"
```

Comment:
Just curious, is this the way PIP reads the metadata ? Is there a way we have multiple metadata files in a wheel such that the metadata says we depend on package A, but in reality when installing via pip, we depend on package B ?

Reply:
Here's how pip finds the .dist-info folder as far as I can tell https://github.com/pypa/pip/blob/bf91a079791f2daf4339115fb39ce7d7e33a9312/src/pip/_internal/utils/wheel.py#L84-L114

Reply:
Maybe upload API should do these checks too. Right now it checks for the `WHEEL` presence and allowed platform tags .

https://github.com/pypa/warehouse/blob/e5cb0e251898e2b60e9d801054d97e12ae3b8e7b/warehouse/forklift/legacy.py#L711-L727

Reply:
> Is there a way we have multiple metadata files in a wheel such that the metadata says we depend on package A, but in reality when installing via pip, we depend on package B ?

According to this code `pip` should fail if there are multiple `.dist-info` dirs
https://github.com/pypa/pip/blob/bf91a079791f2daf4339115fb39ce7d7e33a9312/src/pip/_internal/utils/wheel.py#L100

Reply:
*Technically* a wheel can contain multiple `.dist-info` directories as long as only one matches the name-version pair in the wheel’s file name. But pip adds this additional restriction to simplify implementation, and afaik no-one has complained yet.

Reply:
Those extra `.dist-info` directories will not be used, because [a wheel](https://www.python.org/dev/peps/pep-0427/) ships only one `package-version` build, and the `.dist-info` that will is used should match the `name-version` pair in filename.